File: lib/domain/proposition.dart
import '../data/database.dart';
import 'conflits.dart';
import 'dates.dart';

// Une « cible » : une période à couvrir pour un enfant (un trou non encore assuré).
class Cible {
  final Enfant enfant;
  final DateTime debut;
  final DateTime fin;
  final int besoinId;
  const Cible({
    required this.enfant,
    required this.debut,
    required this.fin,
    required this.besoinId,
  });
}

// Une affectation proposée par le moteur.
class Proposition {
  final Enfant enfant;
  final Accueillant accueillant;
  final DateTime debut;
  final DateTime fin;
  final int besoinId;
  const Proposition({
    required this.enfant,
    required this.accueillant,
    required this.debut,
    required this.fin,
    required this.besoinId,
  });
}

class ResultatProposition {
  final List<Proposition> propositions;
  final List<Cible> nonPlaces;
  final bool limiteAtteinte; // la recherche a été tronquée (problème trop gros)
  const ResultatProposition({
    required this.propositions,
    required this.nonPlaces,
    this.limiteAtteinte = false,
  });
}

// Calcule, sur [debut, fin], les sous-périodes non couvertes par l'union des
// [couvertures] (affectations ET/OU solutions alternatives).
List<(DateTime, DateTime)> trousNonCouverts(
  DateTime debut,
  DateTime fin,
  List<(DateTime, DateTime)> couvertures,
) {
  final res = <(DateTime, DateTime)>[];
  DateTime? debutTrou;
  var d = jour(debut);
  final f = jour(fin);
  while (!d.isAfter(f)) {
    final couvert = couvertures.any(
      (c) => !d.isBefore(jour(c.$1)) && !d.isAfter(jour(c.$2)),
    );
    if (!couvert) {
      debutTrou ??= d;
    } else if (debutTrou != null) {
      res.add((debutTrou, d.subtract(const Duration(days: 1))));
      debutTrou = null;
    }
    d = d.add(const Duration(days: 1));
  }
  if (debutTrou != null) res.add((debutTrou, f));
  return res;
}

// Couvertures (intervalles) d'un enfant : ses affectations + ses solutions.
List<(DateTime, DateTime)> couverturesEnfant(
  int enfantId,
  List<Affectation> affectations,
  List<SolutionAlternative> solutions,
) => [
  for (final a in affectations)
    if (a.enfantId == enfantId && relaisActif(a.statut)) (a.debut, a.fin),
  for (final s in solutions)
    if (s.enfantId == enfantId) (s.debut, s.fin),
];

// Fusionne une liste d'intervalles [debut, fin] (bornes incluses) en intervalles
// disjoints et triés (les périodes qui se chevauchent ou sont contiguës d'un jour
// sont réunies). Sert à éviter de générer deux cibles qui se recoupent pour un
// même enfant lorsque ses besoins se chevauchent.
List<(DateTime, DateTime)> fusionnerPeriodes(List<(DateTime, DateTime)> xs) {
  final valides = [
    for (final (d1, f1) in xs)
      if (!jour(f1).isBefore(jour(d1))) (jour(d1), jour(f1)),
  ]..sort((a, b) => a.$1.compareTo(b.$1));
  if (valides.isEmpty) return const [];
  final res = <(DateTime, DateTime)>[];
  var debut = valides.first.$1;
  var fin = valides.first.$2;
  for (final (d, f) in valides.skip(1)) {
    // Chevauchant ou contigu (le jour suivant la fin courante) : on étend.
    if (!d.isAfter(fin.add(const Duration(days: 1)))) {
      if (f.isAfter(fin)) fin = f;
    } else {
      res.add((debut, fin));
      debut = d;
      fin = f;
    }
  }
  res.add((debut, fin));
  return res;
}

const int _capNoeuds = 200000;

// Propose des affectations couvrant les besoins non assurés, en respectant
// toutes les contraintes bloquantes. Recherche par backtracking avec
// heuristique « variable la plus contrainte d'abord ».
ResultatProposition proposerAffectations({
  required List<Enfant> enfants,
  required List<Accueillant> accueillants,
  required List<Affectation> affectationsExistantes,
  required List<BesoinRelais> besoins,
  required List<DisponibiliteAccueil> dispos,
  required List<Indisponibilite> indispos,
  required List<Incompatibilite> incompatibilites,
  List<Fratrie> fratries = const [],
  List<PreferenceAccueil> preferences = const [],
  List<SolutionAlternative> solutions = const [],
}) {
  final parEnfant = {for (final e in enfants) e.id: e};
  final politiqueParFratrie = {for (final f in fratries) f.id: f.regroupement};
  final dispoParAcc = <int, List<DisponibiliteAccueil>>{};
  for (final d in dispos) {
    (dispoParAcc[d.accueillantId] ??= []).add(d);
  }
  final indispoParAcc = <int, List<Indisponibilite>>{};
  for (final i in indispos) {
    (indispoParAcc[i.accueillantId] ??= []).add(i);
  }
  // Les relais annulés n'occupent plus de place : on les ignore partout.
  final affsActives = affectationsExistantes
      .where((a) => relaisActif(a.statut))
      .toList();

  // 1. Construire les cibles (trous à combler), en fusionnant d'abord les
  //    besoins de chaque enfant pour ne jamais produire deux cibles qui se
  //    recoupent (sinon la règle « déjà placé » en rejetterait une à tort).
  final besoinsParEnfant = <int, List<BesoinRelais>>{};
  for (final b in besoins) {
    if (parEnfant[b.enfantId] == null) continue;
    (besoinsParEnfant[b.enfantId] ??= []).add(b);
  }
  final cibles = <Cible>[];
  for (final entree in besoinsParEnfant.entries) {
    final enfant = parEnfant[entree.key]!;
    final couvertures = couverturesEnfant(
      enfant.id,
      affectationsExistantes,
      solutions,
    );
    final periodes = fusionnerPeriodes([
      for (final b in entree.value) (b.debut, b.fin),
    ]);
    for (final (bDebut, bFin) in periodes) {
      for (final (debut, fin) in trousNonCouverts(bDebut, bFin, couvertures)) {
        // Rattacher le trou au premier besoin qui le chevauche (suivi/besoinId).
        final source = entree.value.firstWhere(
          (b) => periodesSeChevauchent(b.debut, b.fin, debut, fin),
          orElse: () => entree.value.first,
        );
        cibles.add(
          Cible(enfant: enfant, debut: debut, fin: fin, besoinId: source.id),
        );
      }
    }
  }

  if (cibles.isEmpty) {
    return const ResultatProposition(propositions: [], nonPlaces: []);
  }

  // Vérifie qu'une affectation ne crée aucun conflit bloquant, compte tenu des
  // affectations existantes et des propositions déjà retenues.
  bool sansBloquant(Cible c, Accueillant acc, List<Proposition> retenues) {
    final affs = [
      ...affsActives,
      // ids négatifs DISTINCTS pour les propositions retenues : aucune ne peut
      // être confondue avec une vraie affectation (id >= 1), et même si un
      // affectationExclueId était transmis, il n'en exclurait jamais plusieurs.
      for (final (k, p) in retenues.indexed)
        Affectation(
          id: -(k + 1),
          enfantId: p.enfant.id,
          accueillantId: p.accueillant.id,
          debut: p.debut,
          fin: p.fin,
          besoinId: null,
          statut: statutConfirme,
        ),
    ];
    final conflits = analyserAffectation(
      enfant: c.enfant,
      accueillant: acc,
      debut: c.debut,
      fin: c.fin,
      affectations: affs,
      disponibilites: dispoParAcc[acc.id] ?? const [],
      indisponibilites: indispoParAcc[acc.id] ?? const [],
      incompatibilites: incompatibilites,
      enfants: enfants,
      fratries: fratries,
      preferences: preferences,
      solutions: solutions,
    );
    return !conflits.any((x) => x.estBloquant);
  }

  // Candidats classés : fratrie réunie d'abord, puis charge la plus faible.
  List<Accueillant> candidats(Cible c, List<Proposition> retenues) {
    final faisables = accueillants
        .where((a) => sansBloquant(c, a, retenues))
        .toList();
    int scoreFratrie(Accueillant a) {
      if (c.enfant.fratrieId == null) return 0;
      if (politiqueParFratrie[c.enfant.fratrieId] != regroupementEnsemble) {
        return 0;
      }
      final freres = enfants
          .where(
            (e) => e.fratrieId == c.enfant.fratrieId && e.id != c.enfant.id,
          )
          .map((e) => e.id)
          .toSet();
      // Un frère est-il chez cet accueillant sur la période ?
      final viaExistant = affsActives.any(
        (x) =>
            x.accueillantId == a.id &&
            freres.contains(x.enfantId) &&
            periodesSeChevauchent(c.debut, c.fin, x.debut, x.fin),
      );
      final viaRetenu = retenues.any(
        (p) =>
            p.accueillant.id == a.id &&
            freres.contains(p.enfant.id) &&
            periodesSeChevauchent(c.debut, c.fin, p.debut, p.fin),
      );
      return (viaExistant || viaRetenu) ? 1 : 0;
    }

    int charge(Accueillant a) {
      var n = 0;
      for (final x in affsActives) {
        if (x.accueillantId == a.id &&
            periodesSeChevauchent(c.debut, c.fin, x.debut, x.fin)) {
          n++;
        }
      }
      for (final p in retenues) {
        if (p.accueillant.id == a.id &&
            periodesSeChevauchent(c.debut, c.fin, p.debut, p.fin)) {
          n++;
        }
      }
      return n;
    }

    int scoreFavori(Accueillant a) =>
        preferences.any(
          (p) =>
              p.enfantId == c.enfant.id &&
              p.accueillantId == a.id &&
              p.type == prefFavori,
        )
        ? 1
        : 0;

    // Même secteur que l'enfant privilégié (proximité).
    int scoreSecteur(Accueillant a) {
      final sa = a.secteur?.trim().toLowerCase() ?? '';
      final se = c.enfant.secteur?.trim().toLowerCase() ?? '';
      return (sa.isNotEmpty && sa == se) ? 1 : 0;
    }

    // Accueillant encore sous son plafond annuel (sinon dé-priorisé).
    int souPlafond(Accueillant a) {
      final plafond = a.plafondJoursAn;
      if (plafond == null) return 1;
      final annee = jour(c.debut).year;
      var cumul = nbJours(c.debut, c.fin);
      for (final x in affsActives) {
        if (x.accueillantId == a.id && jour(x.debut).year == annee) {
          cumul += nbJours(x.debut, x.fin);
        }
      }
      for (final p in retenues) {
        if (p.accueillant.id == a.id && jour(p.debut).year == annee) {
          cumul += nbJours(p.debut, p.fin);
        }
      }
      return cumul <= plafond ? 1 : 0;
    }

    faisables.sort((a, b) {
      // Favoris d'abord, puis même secteur, fratrie réunie, sous plafond, charge.
      final fav = scoreFavori(b).compareTo(scoreFavori(a));
      if (fav != 0) return fav;
      final sect = scoreSecteur(b).compareTo(scoreSecteur(a));
      if (sect != 0) return sect;
      final f = scoreFratrie(b).compareTo(scoreFratrie(a));
      if (f != 0) return f;
      final pla = souPlafond(b).compareTo(souPlafond(a));
      if (pla != 0) return pla;
      return charge(a).compareTo(charge(b));
    });
    return faisables;
  }

  // Ordonner les cibles par nombre de candidats croissant (les plus contraintes
  // d'abord), évalué sur les seules affectations existantes.
  final ordre = [...cibles];
  ordre.sort(
    (a, b) =>
        candidats(a, const []).length.compareTo(candidats(b, const []).length),
  );

  final courant = <Proposition>[];
  List<Proposition> meilleur = const [];
  var meilleurNb = -1;
  var meilleurFratrie = -1;
  var noeuds = 0;
  var limite = false;

  int bonusFratrie(List<Proposition> sol) {
    var n = 0;
    for (final p in sol) {
      if (p.enfant.fratrieId == null) continue;
      if (politiqueParFratrie[p.enfant.fratrieId] != regroupementEnsemble) {
        continue;
      }
      final freres = enfants
          .where(
            (e) => e.fratrieId == p.enfant.fratrieId && e.id != p.enfant.id,
          )
          .map((e) => e.id)
          .toSet();
      final avecFrere =
          sol.any(
            (q) =>
                q != p &&
                q.accueillant.id == p.accueillant.id &&
                freres.contains(q.enfant.id) &&
                periodesSeChevauchent(p.debut, p.fin, q.debut, q.fin),
          ) ||
          affsActives.any(
            (x) =>
                x.accueillantId == p.accueillant.id &&
                freres.contains(x.enfantId) &&
                periodesSeChevauchent(p.debut, p.fin, x.debut, x.fin),
          );
      if (avecFrere) n++;
    }
    return n;
  }

  void dfs(int i) {
    if (noeuds > _capNoeuds) {
      limite = true;
      return;
    }
    noeuds++;
    // Élagage : impossible de battre le meilleur nombre de placements déjà trouvé.
    final maxPossible = courant.length + (ordre.length - i);
    if (maxPossible < meilleurNb) return;

    if (i == ordre.length) {
      final nb = courant.length;
      final fr = bonusFratrie(courant);
      if (nb > meilleurNb || (nb == meilleurNb && fr > meilleurFratrie)) {
        meilleur = [...courant];
        meilleurNb = nb;
        meilleurFratrie = fr;
      }
      return;
    }

    final c = ordre[i];
    for (final acc in candidats(c, courant)) {
      courant.add(
        Proposition(
          enfant: c.enfant,
          accueillant: acc,
          debut: c.debut,
          fin: c.fin,
          besoinId: c.besoinId,
        ),
      );
      dfs(i + 1);
      courant.removeLast();
      if (noeuds > _capNoeuds) {
        limite = true;
        return;
      }
    }
    // Option : laisser cette cible non placée.
    dfs(i + 1);
  }

  dfs(0);

  // Clé incluant le besoinId : deux cibles d'un même enfant sur les mêmes dates
  // (besoins en double) restent discernables dans le comptage placés/non placés.
  final placesParCle = {
    for (final p in meilleur)
      '${p.besoinId}|${p.enfant.id}|${p.debut}|${p.fin}': p,
  };
  final nonPlaces = ordre
      .where(
        (c) => !placesParCle.containsKey(
          '${c.besoinId}|${c.enfant.id}|${c.debut}|${c.fin}',
        ),
      )
      .toList();

  return ResultatProposition(
    propositions: meilleur,
    nonPlaces: nonPlaces,
    limiteAtteinte: limite,
  );
}

File: web/lib/domain/proposition.ts
// Proposition automatique par backtracking (port de lib/domain/proposition.dart).

import {
  type Accueillant,
  type Affectation,
  type BesoinRelais,
  type DisponibiliteAccueil,
  type Enfant,
  type Fratrie,
  type Incompatibilite,
  type Indisponibilite,
  type PreferenceAccueil,
  type SolutionAlternative,
  prefFavori,
  regroupementEnsemble,
  relaisActif,
  statutConfirme,
} from './types';
import { jour, nbJours, periodesSeChevauchent } from './dates';
import { analyserAffectation, estBloquant } from './conflits';
import { distanceKm } from './distance';

const JOUR_MS = 86400000;

export interface Cible {
  enfant: Enfant;
  debut: Date;
  fin: Date;
  besoinId: number;
}

export interface Proposition {
  enfant: Enfant;
  accueillant: Accueillant;
  debut: Date;
  fin: Date;
  besoinId: number;
}

export interface ResultatProposition {
  propositions: Proposition[];
  nonPlaces: Cible[];
  limiteAtteinte: boolean;
}

type Intervalle = [Date, Date];

export function trousNonCouverts(debut: Date, fin: Date, couvertures: Intervalle[]): Intervalle[] {
  const res: Intervalle[] = [];
  let debutTrou: Date | null = null;
  let d = jour(debut);
  const f = jour(fin);
  while (d.getTime() <= f.getTime()) {
    const couvert = couvertures.some(
      (c) => d.getTime() >= jour(c[0]).getTime() && d.getTime() <= jour(c[1]).getTime(),
    );
    if (!couvert) {
      debutTrou ??= d;
    } else if (debutTrou) {
      res.push([debutTrou, new Date(d.getTime() - JOUR_MS)]);
      debutTrou = null;
    }
    d = new Date(d.getTime() + JOUR_MS);
  }
  if (debutTrou) res.push([debutTrou, f]);
  return res;
}

export function couverturesEnfant(
  enfantId: number,
  affectations: Affectation[],
  solutions: SolutionAlternative[],
): Intervalle[] {
  const res: Intervalle[] = [];
  for (const a of affectations) {
    if (a.enfantId === enfantId && relaisActif(a.statut)) res.push([a.debut, a.fin]);
  }
  for (const s of solutions) {
    if (s.enfantId === enfantId) res.push([s.debut, s.fin]);
  }
  return res;
}

export function fusionnerPeriodes(xs: Intervalle[]): Intervalle[] {
  const valides = xs
    .filter(([d1, f1]) => jour(f1).getTime() >= jour(d1).getTime())
    .map(([d1, f1]) => [jour(d1), jour(f1)] as Intervalle)
    .sort((a, b) => a[0].getTime() - b[0].getTime());
  if (valides.length === 0) return [];
  const res: Intervalle[] = [];
  let [debut, fin] = valides[0];
  for (const [d, f] of valides.slice(1)) {
    if (d.getTime() <= fin.getTime() + JOUR_MS) {
      if (f.getTime() > fin.getTime()) fin = f;
    } else {
      res.push([debut, fin]);
      debut = d;
      fin = f;
    }
  }
  res.push([debut, fin]);
  return res;
}

const CAP_NOEUDS = 200000;

export interface ParamsProposition {
  enfants: Enfant[];
  accueillants: Accueillant[];
  affectationsExistantes: Affectation[];
  besoins: BesoinRelais[];
  dispos: DisponibiliteAccueil[];
  indispos: Indisponibilite[];
  incompatibilites: Incompatibilite[];
  fratries?: Fratrie[];
  preferences?: PreferenceAccueil[];
  solutions?: SolutionAlternative[];
}

export function proposerAffectations(p: ParamsProposition): ResultatProposition {
  const {
    enfants,
    accueillants,
    affectationsExistantes,
    besoins,
    dispos,
    indispos,
    incompatibilites,
    fratries = [],
    preferences = [],
    solutions = [],
  } = p;

  const parEnfant = new Map(enfants.map((e) => [e.id, e]));
  const politiqueParFratrie = new Map(fratries.map((f) => [f.id, f.regroupement]));
  const dispoParAcc = new Map<number, DisponibiliteAccueil[]>();
  for (const d of dispos) {
    (dispoParAcc.get(d.accueillantId) ?? dispoParAcc.set(d.accueillantId, []).get(d.accueillantId)!).push(d);
  }
  const indispoParAcc = new Map<number, Indisponibilite[]>();
  for (const i of indispos) {
    (indispoParAcc.get(i.accueillantId) ?? indispoParAcc.set(i.accueillantId, []).get(i.accueillantId)!).push(i);
  }
  const affsActives = affectationsExistantes.filter((a) => relaisActif(a.statut));

  // 1. Cibles (trous à combler), besoins fusionnés par enfant.
  const besoinsParEnfant = new Map<number, BesoinRelais[]>();
  for (const b of besoins) {
    if (!parEnfant.has(b.enfantId)) continue;
    (besoinsParEnfant.get(b.enfantId) ?? besoinsParEnfant.set(b.enfantId, []).get(b.enfantId)!).push(b);
  }
  const cibles: Cible[] = [];
  for (const [enfantId, liste] of besoinsParEnfant) {
    const enfant = parEnfant.get(enfantId)!;
    const couvertures = couverturesEnfant(enfant.id, affectationsExistantes, solutions);
    const periodes = fusionnerPeriodes(liste.map((b) => [b.debut, b.fin] as Intervalle));
    for (const [bDebut, bFin] of periodes) {
      for (const [debut, fin] of trousNonCouverts(bDebut, bFin, couvertures)) {
        const source =
          liste.find((b) => periodesSeChevauchent(b.debut, b.fin, debut, fin)) ?? liste[0];
        cibles.push({ enfant, debut, fin, besoinId: source.id });
      }
    }
  }

  if (cibles.length === 0) return { propositions: [], nonPlaces: [], limiteAtteinte: false };

  const sansBloquant = (c: Cible, acc: Accueillant, retenues: Proposition[]): boolean => {
    const affs: Affectation[] = [
      ...affsActives,
      ...retenues.map((p2, k) => ({
        id: -(k + 1),
        enfantId: p2.enfant.id,
        accueillantId: p2.accueillant.id,
        debut: p2.debut,
        fin: p2.fin,
        besoinId: null,
        statut: statutConfirme,
        transport: null,
      })),
    ];
    const conflits = analyserAffectation({
      enfant: c.enfant,
      accueillant: acc,
      debut: c.debut,
      fin: c.fin,
      affectations: affs,
      disponibilites: dispoParAcc.get(acc.id) ?? [],
      indisponibilites: indispoParAcc.get(acc.id) ?? [],
      incompatibilites,
      enfants,
      fratries,
      preferences,
      solutions,
    });
    return !conflits.some(estBloquant);
  };

  const candidats = (c: Cible, retenues: Proposition[]): Accueillant[] => {
    const faisables = accueillants.filter((a) => sansBloquant(c, a, retenues));
    const freresDe = (): Set<number> =>
      new Set(
        enfants.filter((e) => e.fratrieId === c.enfant.fratrieId && e.id !== c.enfant.id).map((e) => e.id),
      );
    const scoreFratrie = (a: Accueillant): number => {
      if (c.enfant.fratrieId == null) return 0;
      if (politiqueParFratrie.get(c.enfant.fratrieId) !== regroupementEnsemble) return 0;
      const freres = freresDe();
      const viaExistant = affsActives.some(
        (x) => x.accueillantId === a.id && freres.has(x.enfantId) && periodesSeChevauchent(c.debut, c.fin, x.debut, x.fin),
      );
      const viaRetenu = retenues.some(
        (q) => q.accueillant.id === a.id && freres.has(q.enfant.id) && periodesSeChevauchent(c.debut, c.fin, q.debut, q.fin),
      );
      return viaExistant || viaRetenu ? 1 : 0;
    };
    const charge = (a: Accueillant): number => {
      let n = 0;
      for (const x of affsActives) {
        if (x.accueillantId === a.id && periodesSeChevauchent(c.debut, c.fin, x.debut, x.fin)) n++;
      }
      for (const q of retenues) {
        if (q.accueillant.id === a.id && periodesSeChevauchent(c.debut, c.fin, q.debut, q.fin)) n++;
      }
      return n;
    };
    const scoreFavori = (a: Accueillant): number =>
      preferences.some((x) => x.enfantId === c.enfant.id && x.accueillantId === a.id && x.type === prefFavori) ? 1 : 0;
    const scoreSecteur = (a: Accueillant): number => {
      const sa = (a.secteur ?? '').trim().toLowerCase();
      const se = (c.enfant.secteur ?? '').trim().toLowerCase();
      return sa && sa === se ? 1 : 0;
    };
    const souPlafond = (a: Accueillant): number => {
      if (a.plafondJoursAn == null) return 1;
      const annee = jour(c.debut).getFullYear();
      let cumul = nbJours(c.debut, c.fin);
      for (const x of affsActives) {
        if (x.accueillantId === a.id && jour(x.debut).getFullYear() === annee) cumul += nbJours(x.debut, x.fin);
      }
      for (const q of retenues) {
        if (q.accueillant.id === a.id && jour(q.debut).getFullYear() === annee) cumul += nbJours(q.debut, q.fin);
      }
      return cumul <= a.plafondJoursAn ? 1 : 0;
    };
    faisables.sort((a, b) => {
      const fav = scoreFavori(b) - scoreFavori(a);
      if (fav !== 0) return fav;
      const sect = scoreSecteur(b) - scoreSecteur(a);
      if (sect !== 0) return sect;
      const f = scoreFratrie(b) - scoreFratrie(a);
      if (f !== 0) return f;
      const pla = souPlafond(b) - souPlafond(a);
      if (pla !== 0) return pla;
      // Plus proche d'abord (si les deux adresses sont géolocalisées).
      const da = distanceKm(c.enfant, a);
      const db = distanceKm(c.enfant, b);
      if (da != null && db != null && da !== db) return da - db;
      const ch = charge(a) - charge(b);
      if (ch !== 0) return ch;
      return a.id - b.id; // départage stable (parité avec le desktop)
    });
    return faisables;
  };

  const ordre = [...cibles].sort((a, b) => {
    const n = candidats(a, []).length - candidats(b, []).length;
    if (n !== 0) return n;
    // Départage stable : besoin, puis enfant, puis début (parité desktop).
    if (a.besoinId !== b.besoinId) return a.besoinId - b.besoinId;
    if (a.enfant.id !== b.enfant.id) return a.enfant.id - b.enfant.id;
    return jour(a.debut).getTime() - jour(b.debut).getTime();
  });

  let courant: Proposition[] = [];
  let meilleur: Proposition[] = [];
  let meilleurNb = -1;
  let meilleurFratrie = -1;
  let noeuds = 0;
  let limite = false;

  const bonusFratrie = (sol: Proposition[]): number => {
    let n = 0;
    for (const q of sol) {
      if (q.enfant.fratrieId == null) continue;
      if (politiqueParFratrie.get(q.enfant.fratrieId) !== regroupementEnsemble) continue;
      const freres = new Set(
        enfants.filter((e) => e.fratrieId === q.enfant.fratrieId && e.id !== q.enfant.id).map((e) => e.id),
      );
      const avecFrere =
        sol.some(
          (r) => r !== q && r.accueillant.id === q.accueillant.id && freres.has(r.enfant.id) && periodesSeChevauchent(q.debut, q.fin, r.debut, r.fin),
        ) ||
        affsActives.some(
          (x) => x.accueillantId === q.accueillant.id && freres.has(x.enfantId) && periodesSeChevauchent(q.debut, q.fin, x.debut, x.fin),
        );
      if (avecFrere) n++;
    }
    return n;
  };

  const dfs = (i: number): void => {
    if (noeuds > CAP_NOEUDS) {
      limite = true;
      return;
    }
    noeuds++;
    const maxPossible = courant.length + (ordre.length - i);
    if (maxPossible < meilleurNb) return;
    if (i === ordre.length) {
      const nb = courant.length;
      const fr = bonusFratrie(courant);
      if (nb > meilleurNb || (nb === meilleurNb && fr > meilleurFratrie)) {
        meilleur = [...courant];
        meilleurNb = nb;
        meilleurFratrie = fr;
      }
      return;
    }
    const c = ordre[i];
    for (const acc of candidats(c, courant)) {
      courant.push({ enfant: c.enfant, accueillant: acc, debut: c.debut, fin: c.fin, besoinId: c.besoinId });
      dfs(i + 1);
      courant.pop();
      if (noeuds > CAP_NOEUDS) {
        limite = true;
        return;
      }
    }
    dfs(i + 1);
  };

  dfs(0);

  const cle = (besoinId: number, enfantId: number, debut: Date, fin: Date): string =>
    `${besoinId}|${enfantId}|${debut.getTime()}|${fin.getTime()}`;
  const placees = new Set(meilleur.map((q) => cle(q.besoinId, q.enfant.id, q.debut, q.fin)));
  const nonPlaces = ordre.filter((c) => !placees.has(cle(c.besoinId, c.enfant.id, c.debut, c.fin)));

  return { propositions: meilleur, nonPlaces, limiteAtteinte: limite };
}
